bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -2986,7 +3011,14 @@ void ReplicaImp::onViewsChangeTimer(Timers::Handle timer)  // TODO(GG): review/u
                "Unable to activate the last agreed view (despite receiving 2f+2c+1 view change msgs). "
                "State transfer hasn't kicked-in for a while either. Asking to leave the current view: "
                    << KVLOG(curView, timeSinceLastAgreedViewMilli, timeSinceLastStateTransferMilli));
-      GotoNextView();
+
+      std::unique_ptr<ReplicaAsksToLeaveViewMsg> askToLeaveView(ReplicaAsksToLeaveViewMsg::create(
+          config_.getreplicaId(), curView, ReplicaAsksToLeaveViewMsg::Reason::PrimaryGetInChargeTimeout));
+      sendToAllOtherReplicas(askToLeaveView.get());
+      complainedReplicas.store(std::move(askToLeaveView));
+      metric_sent_replica_asks_to_leave_view_msg_.Get().Inc();
+
+      tryToGotoNextView();
```

Comment:
At some point, shouldn’t we check that we have f+1 complaints before we move to the next view?

Reply:
tryToGotoNextView() does check if we have stored F+1 verified complains (ReplicaAsksToLeaveViewMsg-s) from different Replicas for the current View.
`void ReplicaImp::tryToGotoNextView() {
  if (complainedReplicas.hasQuorumToLeaveView()) {
    GotoNextView();
  } else {
    LOG_INFO(VC_LOG, "Insufficient quorum for moving to next view " << KVLOG(getCurrentView()));
  }
}
bool hasQuorumToLeaveView() const { return msgs.size() >= config_.fVal + 1U; }
`

Reply:
I see, thanks for the clarification!